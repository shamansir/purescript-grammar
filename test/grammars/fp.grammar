main :- expr.

expr :- (funcCall | lambda | letExpr | varExpr | stringLit | intLit | placeholder).

funcCall :- [varExpr, "(", expr, ")"].
parenExpr :- ["(", expr, ")"].
lambda :- ["(", lambdaParam ,")", ws, ":", ws, type, ws, "=>", ws, expr].
lambdaParam :- [ident, ws, ":", ws, type].
letExpr :- [letKW, ws, ident, ws, "=", ws, expr, ws, inKW, ws, expr].
varExpr :- ident.
stringLit :- ["\"", repSep(stringChar, ""), "\""].
intLit :- [num, repSep(num, "")].

type :- ident.

alpha :- ([a-z] | [A-Z] | "_").
num :- [0-9].
alphaNum :- (alpha | num).
ident :- [alpha, repSep((alphaNum | "."), "")].
ws :- repSep((" "|"\n"), "").
placeholder :- "???".
stringChar :- (^'"' | ['\\', '"']).

letKW :- "let".
inKW :- "in".
