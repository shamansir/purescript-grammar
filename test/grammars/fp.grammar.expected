main :- expr.
alpha :- ([a-z]|[A-Z]|"_").
alphaNum :- (alpha|num).
expr :- (funcCall|lambda|letExpr|varExpr|stringLit|intLit|placeholder).
funcCall :- [varExpr,"(",expr,")"].
ident :- [alpha,repSep((alphaNum|"."),"")].
inKW :- "in".
intLit :- [num,repSep(num,"")].
lambda :- ["(",lambdaParam,")",ws,":",ws,type,ws,"=>",ws,expr].
lambdaParam :- [ident,ws,":",ws,type].
letExpr :- [letKW,ws,ident,ws,"=",ws,expr,ws,inKW,ws,expr].
letKW :- "let".
num :- [0-9].
parenExpr :- ["(",expr,")"].
placeholder :- "???".
stringChar :- (^'"'|['\\','"']).
stringLit :- ["\"",repSep(stringChar,""),"\""].
type :- ident.
varExpr :- ident.
ws :- repSep((" "|"\n"),"").